src/Scout/ScoutEngine.php

```diff
@@ -0,0 +1,540 @@
+        $cursor->setTypeMap(self::TYPEMAP);
+
+        return $cursor->toArray();
+    }
+
+    /**
+     * Pluck and return the primary keys of the given results.
+     *
+     * @see Engine::mapIds()
+     *
+     * @param list<array|object> $results
+     */
+    #[Override]
+    public function mapIds($results): Collection
+    {
+        assert(is_array($results), new TypeError(sprintf('Argument #1 ($results) must be of type array, %s given', get_debug_type($results))));
+
+        return new Collection(array_column($results, '_id'));
+    }
+
+    /**
+     * Map the given results to instances of the given model.
+     *
+     * @see Engine::map()
+     *
+     * @param Builder $builder
+     * @param array   $results
+     * @param Model   $model
+     *
+     * @return Collection
+     */
+    #[Override]
+    public function map(Builder $builder, $results, $model): Collection
+    {
+        return $this->performMap($builder, $results, $model, false);
+    }
+
+    /**
+     * Map the given results to instances of the given model via a lazy collection.
+     *
+     * @see Engine::lazyMap()
+     *
+     * @param Builder $builder
+     * @param array   $results
+     * @param Model   $model
+     *
+     * @return LazyCollection
+     */
+    #[Override]
+    public function lazyMap(Builder $builder, $results, $model): LazyCollection
+    {
+        return $this->performMap($builder, $results, $model, true);
+    }
+
+    /** @return ($lazy is true ? LazyCollection : Collection)<mixed> */
+    private function performMap(Builder $builder, array $results, Model $model, bool $lazy): Collection|LazyCollection
+    {
+        if (! $results) {
+            $collection = $model->newCollection();
+
+            return $lazy ? LazyCollection::make($collection) : $collection;
+        }
+
+        $objectIds = array_column($results, '_id');
+        $objectIdPositions = array_flip($objectIds);
+
+        return $model->queryScoutModelsByIds($builder, $objectIds)
+            ->{$lazy ? 'cursor' : 'get'}()
+            ->filter(function ($model) use ($objectIds) {
+                return in_array($model->getScoutKey(), $objectIds);
+            })
+            ->map(function ($model) use ($results, $objectIdPositions) {
+                $result = $results[$objectIdPositions[$model->getScoutKey()]] ?? [];
+
+                foreach ($result as $key => $value) {
+                    if ($key[0] === '_' && $key !== '_id') {
+                        $model->withScoutMetadata($key, $value);
+                    }
+                }
+
+                return $model;
+            })
+            ->sortBy(function ($model) use ($objectIdPositions) {
+                return $objectIdPositions[$model->getScoutKey()];
+            })
+            ->values();
+    }
+
+    /**
+     * Get the total count from a raw result returned by the engine.
+     * This is an estimate if the count is larger than 1000.
```

Comment:
A `@see` reference to https://www.mongodb.com/docs/atlas/atlas-search/counting/ may be helpful here. You can also mention that this pertains to the pipeline created in `performSearch()` since that's what ultimately determines this.